### ISSUE TITLE
Tests needs to clean after themselves

### ISSUE THREAD

**Issue by caglar10ur** [vmware/vic#1091]
Otherwise we started to get resource errors like during CI run

```
vm_test.go:146: ERROR: Operation failed: Failed to extend swap file from 0 KB to 2097152 KB. - (&{{} %!s(*types.GenericVmConfigFault=&{{{{<nil> [{{} vob.swap.extend.failure.status [{{} 1 0} {{} 2 2097152}] Failed to extend swap file from 0 KB to 2097152 KB.} {{} vob.swap.extend.failure.moreinfo [{{} 1 0}] Current swap file size is 0 KB.} {{} msg.vmmonVMK.creatVMFailed [{{} 1 msg.vmk.status.VMK_NO_SPACE}] Could not power on virtual machine: No space left on device.} {{} msg.monitorLoop.createVMFailed.vmk [] Failed to power on VM.} {{} msg.vmx.poweron.failed [] Failed to start the virtual machine.}]}}} Failed to extend swap file from 0 KB to 2097152 KB.}) Failed to extend swap file from 0 KB to 2097152 KB.})
```

**Comment by mhagen-vmware** [comment]
Alright, few areas I have been able to identify tonight:
1. Unit tests leave a deadbeef folder hanging around.  It doesn't really take any space, but we should clean it up.  (Carefully so as to maintain multi-tenancy support) * this can probably be very low priority.
2. When E2E tests are abruptly killed for various reasons, any number of containers or VCH appliances could be left hanging around in various powered/indeterminate states.
3. The new docker integration tests on a successful run, likely due to our current partial support, leave multiple containers in various states of power which also prevents us from removing the VCH and cleaning the datastore.
4. I already started looking at it today, but I need to finish the work of cleaning the VIC folder without wiping it out completely to allow for multi-tenancy testing.

4. should provide major savings, and it is the most straightforward to implement.  3. is likely manageable with some work, since we know which VCH appliance is associated with the docker integration tests, as long as we can walk down which containers are associated with it then it should be straightforward to ensure they are powered down and deleted to ultimately cleanup the VCH. 2. is more problematic, need to think on it some more...

**Comment by fdawg4l** [comment]
@mhagen-vmware what do you mean by "multi-tenancy"?  Are you co-locating test infra with dev infra or are you running tests in parallel on the same infra?

**Comment by mhagen-vmware** [comment]
Yea, sorry - the former: multiple tests running in parallel.

**Comment by fdawg4l** [comment]
Got it.  Thanks @mhagen-vmware!  

@caglar10ur and I discovered an issue in the storage layer where using the same host for multiple VCHs will race on a statefile (called the `parentMap`).  I'll fix this today so each VCH on a host has it's own parentMap file.  If you haven't seen issues with this, then I wouldn't worry about it.  In any case, it'll be taken care of ASAP.

**Comment by mhagen-vmware** [comment]
Alright number 4 should be taken care of with #1094 

I have taken a first pass at number 3 with https://github.com/vmware/vic/commit/ca6567d203eac5a6552a826c9cd0cdeaeda59fa1

My initial idea for Number 2 is time-based.  I can pull the creation date of all VMs on the cluster.  Anything older than N should be invalid.  N could be our total test time + buffer, something like 2hrs... but I don't think we even need to be that lenient, containers should not hang around for more than a few minutes during tests and any VCH appliance should only hang around for our longest single test time.  I am thinking 1hr is pretty fair.  Thoughts?

**Comment by caglar10ur** [comment]
for number 2 - I think it could easier if you design the tests for clean startup instead of clean exit. Eg; check the env to see whether you have left overs, remove them, start testing instead of trying to exit cleanly.

**Comment by mhagen-vmware** [comment]
Yep, that was what I was planning for number 2, but we still have the same problem of detecting which ones to delete and which ones might still be active regardless of when we start the check.  Any thoughts on time comparison or better approach?

**Comment by mhagen-vmware** [comment]
Alright, looks a lot better now with: https://github.com/vmware/vic/pull/1100

But something is not cleaning up the datastore still and I am leaning towards our unit tests.  Something is leaving behind quite a few containers.  See #1106 

That and another container running /bin/cat which could also be one of our unit tests?  I can't really find any bread crumb trail as to who is creating them just yet beyond the log file I put in #1106 which really doesn't look like it tells us much other than the command trying to be run.

**Comment by hickeng** [comment]
Failure due to existing directory: https://ci.vmware.run/vmware/vic/1960

Using a static name is fundamentally broken if we're potentially running multiple tests concurrently on the same infra - we can add the build id into the names perhaps if that's available in the environment.

**Comment by mhagen-vmware** [comment]
Fixed by all the previous issues and now: #1270
